### ISSUE TITLE
macos binary relocation not working

### ISSUE THREAD

**Issue by eugeneswalker** [LLNL/spack#32571]
### Steps to reproduce

1. Build package in long path
```
$> spack config add config:install_tree:padded_length:64

$> spack install pkgconf
==> Installing gnuconfig-2021-08-14-6ge5mev6dobcylwn3g5yrkqh4ltosyjj
==> No binary for gnuconfig-2021-08-14-6ge5mev6dobcylwn3g5yrkqh4ltosyjj found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/69/69b6d2868e70167ba1bdb9030b49beeb20f00b37e30825e83fd04291d96bc5f7.tar.gz
==> No patches needed for gnuconfig
==> gnuconfig: Executing phase: 'install'
==> gnuconfig: Successfully installed gnuconfig-2021-08-14-6ge5mev6dobcylwn3g5yrkqh4ltosyjj
  Fetch: 0.07s.  Build: 0.14s.  Total: 0.21s.
[+] /Users/gitlab-runner/spack/opt/spack/[padded-to-63-chars]/darwin-monterey-m1/apple-clang-13.1.6/gnuconfig-2021-08-14-6ge5mev6dobcylwn3g5yrkqh4ltosyjj
==> Installing pkgconf-1.8.0-ett7ygi7ft3lg6iekce3ts2qscv7kxk2
==> No binary for pkgconf-1.8.0-ett7ygi7ft3lg6iekce3ts2qscv7kxk2 found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/ef/ef9c7e61822b7cb8356e6e9e1dca58d9556f3200d78acab35e4347e9d4c2bbaf.tar.xz
==> No patches needed for pkgconf
==> pkgconf: Executing phase: 'autoreconf'
==> pkgconf: Executing phase: 'configure'
==> pkgconf: Executing phase: 'build'
==> pkgconf: Executing phase: 'install'
==> pkgconf: Successfully installed pkgconf-1.8.0-ett7ygi7ft3lg6iekce3ts2qscv7kxk2
  Fetch: 0.08s.  Build: 5.52s.  Total: 5.59s.
[+] /Users/gitlab-runner/spack/opt/spack/[padded-to-63-chars]/darwin-monterey-m1/apple-clang-13.1.6/pkgconf-1.8.0-ett7ygi7ft3lg6iekce3ts2qscv7kxk2
```

2. Create binary cache
```
$> spack buildcache create -afru -d ~ pkgconf
```

3. Install from binary cache into a short path (thus requiring relocation)
```
$> spack config rm config:install_tree:padded_length:64
$> spack mirror add homedir ~
$> spack install --cache-only --no-check-signature pkgconf
[+] /Users/gitlab-runner/spack/opt/spack/darwin-monterey-m1/apple-clang-13.1.6/pkgconf-1.8.0-ett7ygi7ft3lg6iekce3ts2qscv7kxk2
```

4. Package doesn't work
```
$> spack load pkgconf
$> pkgconf
zsh: killed     pkgconf

$> $(spack location -i pkgconf)/bin/pkgconf
dyld[74653]: Library not loaded: @loader_path/../../../../../spackdarwin-monterey-m1/apple-clang-13.1.6/pkgconf-1.8.0-ett7ygi7ft3lg6iekce3ts2qscv7kxk2/lib/libpkgconf.3.dylib
  Referenced from: /Users/gitlab-runner/spack/opt/spack/darwin-monterey-m1/apple-clang-13.1.6/pkgconf-1.8.0-ett7ygi7ft3lg6iekce3ts2qscv7kxk2/bin/pkgconf
  Reason: tried: '/Users/gitlab-runner/spack/opt/spack/darwin-monterey-m1/apple-clang-13.1.6/pkgconf-1.8.0-ett7ygi7ft3lg6iekce3ts2qscv7kxk2/bin/../../../../../spackdarwin-monterey-m1/apple-clang-13.1.6/pkgconf-1.8.0-ett7ygi7ft3lg6iekce3ts2qscv7kxk2/lib/libpkgconf.3.dylib' (no such file), '/libpkgconf.3.dylib' (no such file)
zsh: abort
```

Installing from cache works as long as no relocation involved:
```
$> spack config add config:install_tree:padded_length:64
$> spack uninstall --all
$> spack install --cache-only --no-check-signature pkgconf
$> spack load pkgconf
$> pkgconf
Please specify at least one package name on the command line.
```

FYI @scottwittenburg @adamjstewart @wspear @gartung 

### Error message

See above

### Information on your system

* **Spack:** 0.19.0.dev0 (762ba27036e717aeceeb1b43f3090f0e9a049869)
* **Python:** 3.9.13
* **Platform:** darwin-monterey-m1
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by gartung** [comment]
Ask @nhanford

**Comment by adamjstewart** [comment]
Is anyone working on this? This seems like a major issue.

**Comment by tgamblin** [comment]
@nhanford can you look at this?

**Comment by haampie** [comment]
fwiw, can't reproduce on catalina x86_64 w/ apple-clang 11, seems like it's using absolute paths for load commands and dylib ids instead of `@loader_path`

**Comment by adamjstewart** [comment]
I can reproduce this on my system as well:

* **Spack:** 0.20.0.dev0 (fe995542ab615ad16e13e1143ea3d5a5b5a0eb94)
* **Python:** 3.10.8
* **Platform:** darwin-monterey-m1
* **Concretizer:** clingo

**Comment by alalazo** [comment]
This one is very likely an artifact of mandatory code signing on Apple M1. If relocation tampers the binary, besides RPATH, the "Trusted execution system" kills it before starting - see e.g. https://developer.apple.com/forums/thread/698167

**Comment by adamjstewart** [comment]
@bernhardkaindl mentioned something about this on Slack, maybe he could work on this? @haampie offered to work on this if we can get him access to an M1 system.

**Comment by eugeneswalker** [comment]
> @bernhardkaindl mentioned something about this on Slack, maybe he could work on this? @haampie offered to work on this if we can get him access to an M1 system.

We can get @haampie an account on our M1.

**Comment by eugeneswalker** [comment]
Just checked a new M2 Mac we have, running `macOS 13.0 Ventura`, problem exists out of the box there too:

* **Spack:** 0.20.0.dev0 (3695200a3cba4a066e63d8318755b88106e11499)
* **Python:** 3.9.6
* **Platform:** darwin-ventura-m2
* **Concretizer:** clingo

I think it is this trusted build issue. I'm trying to figure out how to disable this or how we can work around this.

**Comment by haampie** [comment]
Should be fixed by #35585

**Comment by eugeneswalker** [comment]
> Should be fixed by #35585

Thanks!
